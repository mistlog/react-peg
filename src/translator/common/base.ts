import { Chunk } from "../../chunk/chunk";

export abstract class Translator {
    abstract translate(chunk: Chunk, actions: Map<string, Function>): string;
}