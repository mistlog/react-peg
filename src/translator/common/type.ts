import { RepeatType, AssertType } from "../../chunk/chunk";

export interface BuiltInRule {
    label?: string;
}

export interface TextRule extends BuiltInRule {
    children?: string;
}

export interface AnyCharacterRule extends BuiltInRule {
    
}

export interface OrRule extends BuiltInRule {
    children: Array<JSX.Element>;
}

export interface ListRule extends BuiltInRule {
    children: JSX.Element | Array<JSX.Element>;
}

export interface RepeatRule extends BuiltInRule {
    children: JSX.Element | Array<JSX.Element>;
    type: RepeatType;
}

export interface SetRule extends BuiltInRule {
    children: string;
}

export interface OptionalRule extends BuiltInRule {
    children: JSX.Element | Array<JSX.Element>;
}

export interface AssertRule extends BuiltInRule {
    children: JSX.Element;
    type: AssertType;
}

export interface PatternRule extends BuiltInRule {
    children: JSX.Element | Array<JSX.Element>;
    action: Function;
}