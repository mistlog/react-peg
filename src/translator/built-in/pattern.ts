import { Translator } from "../common/base";
import { Chunk, isChunk } from "../../chunk/chunk";
import { translateChildren } from "../common/pattern";

export class PatternTranslator extends Translator {
    translate(chunk: Chunk, actions: Map<string, Function>) {
        const expression = translateChildren(chunk, actions).join(" ");
        const labels = collectLabels(chunk);
        const actionStr = translateAction(chunk.props["action"], labels, actions);
        return `(${expression} ${actionStr})`;
    }
}

function translateAction(action: Function, labels: Array<string>, actions: Map<string, Function>) {
    const labels_param = labels.length === 0 ? "" : `${labels.join(", ")}, `;

    const key = `action${actions.size}`;
    actions.set(key, action);

    const actionStr = `{ 
        return actions.get("${key}")({${labels_param}${["globalFunction", "globalContext"].join(",")}});
    }`;
    return actionStr;
}

function collectLabels(chunk: Chunk): Array<string> {
    if (!isChunk(chunk)) {
        return [];
    }

    const labels = [];

    if (chunk.props.label) {
        labels.push(chunk.props.label);
    }
    else {
        chunk.children.forEach(child => {
            labels.push(...collectLabels(child as Chunk));
        });
    }

    return labels;
}

