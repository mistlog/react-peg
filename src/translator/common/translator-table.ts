import { TextTranslator } from "../built-in/text";
import { RuleTranslator } from "../rule/rule";
import { OrTranslator } from "../built-in/or";
import { ListTranslator } from "../built-in/list";
import { RepeatTranslator } from "../built-in/repeat";
import { OptionalTranslator } from "../built-in/opt";
import { AssertTranslator } from "../built-in/assert";
import { SetTranslator } from "../built-in/set";
import { PatternTranslator } from "../built-in/pattern";
import { AnyCharacterTranslator } from "../built-in/any";

export const TranslatorTable =
{
    "rule": RuleTranslator,
    "list": ListTranslator,
    "repeat": RepeatTranslator,
    "set": SetTranslator,
    "opt": OptionalTranslator,
    "assert": AssertTranslator,
    "text": TextTranslator,
    "or": OrTranslator,
    "pattern": PatternTranslator,
    "any": AnyCharacterTranslator
}