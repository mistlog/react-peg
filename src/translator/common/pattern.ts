import { Chunk } from "../../chunk/chunk";
import { TranslatorTable } from "./translator-table";

export function translate(chunk: Chunk, actions: Map<string, Function> = new Map<string, Function>()) {
    const label = chunk.props.label ? `${chunk.props.label}: ` : "";
    let pattern = "__undefined__";

    if (chunk.tagName) {
        const key = chunk.tagName;
        const Translator = TranslatorTable[key];
        pattern = new Translator().translate(chunk, actions);
    }
    else {
        pattern = chunk.rule.name;
    }

    return `${label}${pattern}`;
}

export function translateChildren(chunk: Chunk, actions: Map<string, Function>) {
    return (chunk.children as Array<Chunk>).map(child => translate(child, actions));
}