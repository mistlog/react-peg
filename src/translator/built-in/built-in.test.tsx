import { ReactPeg } from "../../index";
import { translate } from "../common/pattern";

test("translator.text", () => {
    const chunk = <text>abc</text>;
    const pattern = translate(chunk);
    expect(pattern).toEqual(`'abc'`);
})

test("translator.or", () => {
    const chunk = (
        <or>
            <text>a</text>
            <text>b</text>
        </or>
    );
    const pattern = translate(chunk);
    expect(pattern).toEqual(`('a' / 'b')`);
})

test("translator.list", () => {
    const chunk = (
        <list>
            <text>a</text>
            <text>b</text>
        </list>
    );
    const pattern = translate(chunk);
    expect(pattern).toEqual(`'a' 'b'`);
})

test("translator.repeat.zero-or-more", () => {
    const chunk = (
        <repeat type="*">
            <text>a</text>
        </repeat>
    );

    const pattern = translate(chunk);
    expect(pattern).toEqual("('a')*");
})

test("translator.repeat.one-or-more", () => {
    const chunk = (
        <repeat type="+">
            <text>a</text>
        </repeat>
    );

    const pattern = translate(chunk);
    expect(pattern).toEqual("('a')+");
});

test("translator.set", () => {
    const chunk = (
        <set>a-z0-9A-Z</set>
    );

    const pattern = translate(chunk);
    expect(pattern).toEqual("[a-z0-9A-Z]");
})

test("translator.optional", () => {
    const chunk = (
        <list>
            <opt>
                <text>a</text>
            </opt>
            <text>a</text>
        </list>
    );
    const pattern = translate(chunk);
    expect(pattern).toEqual(`('a')? 'a'`);
})

test("translator.assert.with", () => {
    const chunk = (
        <list>
            <text>a</text>
            <assert type="with">
                <text>cd</text>
            </assert>
            <text>cdab</text>
        </list>
    );
    const pattern = translate(chunk);
    expect(pattern).toEqual(`'a' &('cd') 'cdab'`);
})

test("translator.assert.without", () => {
    const chunk = (
        <list>
            <text>a</text>
            <assert type="without">
                <text>cd</text>
            </assert>
            <text>cdab</text>
        </list>
    );
    const pattern = translate(chunk);
    expect(pattern).toEqual(`'a' !('cd') 'cdab'`);
})

test("translator.text.label", () => {
    const chunk = <text label="t">abc</text>;
    const pattern = translate(chunk);
    expect(pattern).toEqual(`t: 'abc'`);
})


test("translator.pattern", () => {
    const chunk = (
        <or>
            <pattern action={({ a }) => a + "a"}>
                <text label="a">a</text>
            </pattern>
            <pattern action={({ b }) => b + "b"}>
                <text label="b">b</text>
            </pattern>
        </or>
    );
    const actions = new Map<string, Function>();
    const pattern = translate(chunk, actions);
    expect(pattern).toMatchSnapshot();
})