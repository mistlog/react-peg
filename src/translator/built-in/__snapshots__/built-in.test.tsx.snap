// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translator.pattern 1`] = `
"((a: 'a' { 
        return actions.get(\\"action0\\")({a, globalFunction,globalContext});
    }) / (b: 'b' { 
        return actions.get(\\"action1\\")({b, globalFunction,globalContext});
    }))"
`;
